core/pkg/artifacts/file_cache.go

```diff
@@ -0,0 +1,147 @@
+package artifacts
+
+import (
+	"crypto/md5"
+	"encoding/base64"
+	"io"
+	"log/slog"
+	"os"
+	"path/filepath"
+
+	atomicfile "github.com/natefinch/atomic"
+	"github.com/wandb/wandb/core/pkg/utils"
+)
+
+type FileCache struct {
+	root    string
+	enabled bool
+}
+
+func NewFileCache(enabled bool) FileCache {
+	if !enabled {
+		return FileCache{enabled: false}
+	}
+	dir, found := os.LookupEnv("WANDB_CACHE_DIR")
+	if !found {
+		userCacheDir, err := os.UserCacheDir()
+		if err != nil {
+			slog.Error("Unable to find cache directory", "err", err)
+			return FileCache{enabled: false}
+		}
+		dir = filepath.Join(userCacheDir, "wandb")
+	}
+	return FileCache{root: filepath.Join(dir, "artifacts"), enabled: true}
+}
+
+// AddFile copies a file into the cache and returns the B64MD5 cache key.
+func (c *FileCache) AddFile(path string) (string, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+	return c.Write(f)
+}
+
+// AddFileAndCheckDigest copies a file into the cache. If a digest is provided, it also
+// verifies that the file's MD5 hash matches the digest.
+//
+// This is a convenience function intended for running in the background; instead of
+// returning an error it just logs it. This is often correct, since failing to add a
+// file to the cache shouldn't be a critical error.
+func (c *FileCache) AddFileAndCheckDigest(path string, digest string) {
+	if !c.enabled {
```

Comment:
small nit, I see a lot of these check, might be worthwhile to have `NoOpFileCache` and return that instead in the constructor if needed